### PR TITLE
Allow systems to specify custom filter for the applicable set beyond required_components

### PR DESCRIPTION
**Pull request by Wumpf**
### What

Part of general query effort, mostly
* [#4388](https://github.com/rerun-io/rerun/issues/4388)

Allows visualizers to specify `VisualizerAdditionalApplicabilityFilter` which can be used to answer global visualizer applicability questions like "is this an image tensor". Therefore moving these kind of checks out of the `heuristic_filter`.

Right now this has a negligible (presumed positive) impact on performance. However, there's two strong motivating factors here:
* `heuristic_filter` is right now incorrectly applied on a per space view _class_ bases. It currently is used to extract the "visualizable set" which actually **has** to be determined per space view _instance_ since it is dependent on things like the space's origin. Once `heuristic_filter` runs per instance (or much much worse, per instance candidate during heuristic!) the now moved checks become a huge burden. This is about to happen in a follow-up PR!
* previously, we did a latest-at-query for things that should be resolved with a store diff. Therefore, this change makes the entity->visualizer mapping more correct & deterministic since missing an update because of rendering pacing is no longer possible!

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using newly built examples: [app.rerun.io](https://app.rerun.io/pr/4604/index.html)
  * Using examples from latest `main` build: [app.rerun.io](https://app.rerun.io/pr/4604/index.html?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [app.rerun.io](https://app.rerun.io/pr/4604/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4604)
- [Docs preview](https://rerun.io/preview/a5b72a18a7c05938ec66e510aea455fc42b3608b/docs) 
- [Examples preview](https://rerun.io/preview/a5b72a18a7c05938ec66e510aea455fc42b3608b/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: crates/re_space_view_bar_chart/src/view_part_system.rs [hunk 1]
```diff
@@ -25,6 +25,19 @@ impl IdentifiedViewSystem for BarChartViewPartSystem {
...
 
+struct BarChartVisualizerEntityFilter;
+
+impl VisualizerAdditionalApplicabilityFilter for BarChartVisualizerEntityFilter {
+    fn update_applicability(&mut self, event: &re_arrow_store::StoreEvent) -> bool {
+        event.diff.cells.iter().any(|(component_name, cell)| {
+            component_name == &re_types::components::TensorData::name()
+                && re_types::components::TensorData::from_arrow(cell.as_arrow_ref())
+                    .map(|tensors| tensors.iter().any(|tensor| tensor.is_vector()))
+                    .unwrap_or(false)
```
Comment (jleibs) [inline crates/re_space_view_bar_chart/src/view_part_system.rs:36]: this seems likely to be a recurring pattern we'll want a helper for at some point.

File: crates/re_space_view_spatial/src/parts/images.rs [hunk 2]
```diff
@@ -643,6 +643,23 @@ impl IdentifiedViewSystem for ImagesPart {
...
 }
 
+struct ImageVisualizerEntityFilter;
+
+impl VisualizerAdditionalApplicabilityFilter for ImageVisualizerEntityFilter {
+    fn update_applicability(&mut self, event: &re_arrow_store::StoreEvent) -> bool {
+        event.diff.cells.iter().any(|(component_name, cell)| {
+            component_name == &re_types::components::TensorData::name()
+                && re_types::components::TensorData::from_arrow(cell.as_arrow_ref())
+                    .map(|tensors| {
```
Comment (jleibs) [inline crates/re_space_view_spatial/src/parts/images.rs:653]: yup... there it is again :sweat_smile:

**Review by jleibs — APPROVED** [review]
Nice -- I like how easy it is to add something like this now!

File: crates/re_space_view_bar_chart/src/view_part_system.rs (hunk 1 above)
Comment (Wumpf) [reply crates/re_space_view_bar_chart/src/view_part_system.rs:36]: adding some form of utility that's generic over the component type!

**Comment by Wumpf** [comment]
linkinator failure was temp docker unavailablility

**Pull request merged**